libs/ledgerjs/packages/react-native-hw-transport-ble/src/BleTransport.test.ts

```diff
@@ -0,0 +1,278 @@
+                          break;
+                        // just used for a non resolving apdu
+                        case "0500000005b020000000":
+                          setTimeout(() => {
+                            callbacks?.onDeviceResponse(null, {
+                              value: Buffer.from("05000000029000", "hex"),
+                            });
+                          }, 600);
+                          return; // Called after a delay to give time for the disconnect
+                        default:
+                          throw new Error("some generic failure");
+                      }
+                      // Introduce some logic to actually respond.
+                      callbacks?.onDeviceResponse(null, {
+                        value,
+                      });
+                    },
+                  },
+                ];
+              }
+              throw Error("Generic mocked error");
+            },
+          };
+        },
+      };
+    },
+  };
+});
+
+describe("BleTransport connectivity test coverage", () => {
+  const deviceId = "20EDD96F-7430-6E33-AB22-DD8AAB857CD4";
+
+  describe("Device available and already paired", () => {
+    it("should find the device, connect, negotiate MTU", async () => {
+      const transport = await BleTransport.open(deviceId);
+      expect(transport.device.isConnected()).toBe(true);
+    });
+
+    it("should be disconnectable, and cleanup", async () => {
+      const transport = await BleTransport.open(deviceId);
+      await BleTransport.disconnect(deviceId);
+      expect(transport.isConnected).toBe(false);
+    });
+
+    it("should disconnect in 1s (5s default) after calling close", async () => {
+      const transport = await BleTransport.open(deviceId);
+      expect(transport.isConnected).toBe(true);
+
+      BleTransport.disconnectTimeoutMs = 500;
+      await transport.close();
+
+      // Expect the timeout for disconnection to be set
+      expect(transport.disconnectTimeout).not.toBe(undefined);
+      let resolve;
+
+      transport.on("disconnect", () => {
+        resolve();
+      });
+
+      return await new Promise((_resolve, _reject) => {
+        resolve = _resolve;
+      });
+    });
+
+    it("should cancel disconnect if new connection is made", async () => {
+      const transport = await BleTransport.open(deviceId);
+      expect(transport.isConnected).toBe(true);
+
+      BleTransport.disconnectTimeoutMs = 500;
+      await transport.close();
+
+      // Expect the timeout for disconnection to be set
+      expect(transport.disconnectTimeout).not.toBe(undefined);
+      // Nb due to the different environments, the timeout behaves differently here
+      // and I can't check against a number for it to be cleared or not.
+      expect((transport.disconnectTimeout as any)._destroyed).toBe(false);
+      await BleTransport.open(deviceId);
+      expect((transport.disconnectTimeout as any)._destroyed).toBe(true);
```

Comment:
it's working so it's good

but next time you can try to play with [jest fake timer](https://jestjs.io/docs/timer-mocks) and its `jest.advanceTimersByTime(<time in ms>)`, maybe it would work

Reply:
It wasn't working, some timer somewhere wasn't getting mocked and I gave up